Add airflowctl codegen

aritra24 opened the pull request:
closes: [#47509](https://github.com/apache/airflow/issues/47509)
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

aritra24 committed:
Add airflowctl codegen

aritra24 commented on airflow-ctl/src/airflowctl/api/datamodels/generated.py:
```diff
@@ -136,7 +129,7 @@ class BulkDeleteActionConnectionBody(BaseModel):
         list[str], Field(description="A list of entity id/key to be deleted.", title="Entities")
     ]
-    action_on_non_existence: BulkActionNotOnExistence | None = "fail"
+    action_on_non_existence: BulkActionNotOnExistence | None = BulkActionNotOnExistence.FAIL
```
I made these changes by hand, I wasn't sure why the mypy was failing, I guess because `# mypy: disable-error-code="assignment"` this line was removed.

bugraoz93 replied:
Seems like it. I think we can exclude it for now since it is not a thing we are implementing and can find an automated way as a next step
Edit: exclude from mypy checks

potiuk replied:
@aritra24 @bugraoz93 -> it's perfectly fine to have no licence in generated code.

 We just have to exclude such generated files in .rat-excludes and in the pre-commits that are adding licence.

aritra24 replied:
Oh... 🥲. I think we do need the script anyways though cos we need to add the mypy ignore line so might as well also add the license? What's your opinion?

potiuk replied:
We can just ignore all mypy errors comingfrom this package - we do it in few other generated packages.

potiuk replied:
See pyproject.toml `ignore_errors`

potiuk replied:
Why? Such modificatiosn of generated code are very, very prone to future errors - codegen generation might change any time and someone will have to dig it out and find out how to modify the resuling code to fix the issue - and since that will happen randomly and usually will break someone's PR, that might be super confusing.

bugraoz93 replied:
@potiuk Thanks! I am okay with excluding the license. I saw we have one in `task-sdk`. That's why I was thinking maybe we can do something similar here. 

@aritra24, I agree with Jarek that this package will break at some point without us making any changes. Let's use `ignore_errors` for this package and ignore everything rather than chasing and fixing specific rules. There won't be any code piece under this package other than generated Pydantic objects.

aritra24 replied:
Ack, that's fair. I'll update it to just use the regular code gen. Also if you have any thoughts on the usage of the yaml file as input instead of the api server let me know. 🤔

aritra24 committed:
Added the generated file

bugraoz93 commented on .pre-commit-config.yaml:
```diff
@@ -1345,6 +1345,13 @@ repos:
+        language: python
+        entry: uv run -p 3.12 --no-dev --no-progress --active --group codegen --project apache-airflow-ctl --directory airflow-ctl/ datamodel-codegen
+        pass_filenames: false
+        files: ^airflow-core/src/airflow/api_fastapi/execution_api/.*\.py$
```
```suggestion
        files: ^airflow-core/src/airflow/api_fastapi/core_api/.*\.py$
```

aritra24 replied:
Ack

bugraoz93 reviewed (commented):
Looks good. Thanks!
We may need something similar to this.
https://github.com/apache/airflow/blob/main/task-sdk/dev/datamodel_code_formatter.py
Also, we should do the same automation for auth endpoints where there are different openapi specs and endpoints

aritra24 committed:
Adds license generated.py

Moved datamodel_code_formatter to devel-common
and updated pyproject.toml

aritra24 committed:
Merge branch 'main' into generate_models

aritra24 committed:
Create simple extensible formatter

Currently only takes mypy ignores to append
to the file. Unfortunately datamodel-codegen
only supports a single string so would have to
put in all the ignores into one string if there's
multiple.

bugraoz93 commented on airflow-ctl/pyproject.toml:
```diff
@@ -126,8 +127,11 @@ use-schema-description=true  # Desc becomes class doc comment
 use-standard-collections=true # list[] not List[]
 use-subclass-enum=true # enum, not union of Literals
 use-union-operator=true # 3.9+annotations, not `Union[]`
+custom-formatters = 'misc.simple_code_formatter'
```
Let's update the comments above, too if anyone wants to run manually in local

bugraoz93 replied:
Also there is a TODO we can also delete that

bugraoz93 commented on task-sdk/dev/generate_task_sdk_models.py:
```diff
@@ -40,7 +40,6 @@
 sys.path.insert(0, AIRFLOW_CORE_SOURCES_PATH.as_posix())
 sys.path.insert(0, str(AIRFLOW_ROOT_PATH))  # make sure setup is imported from Airflow
 
-
```
Is this intentional? :)

aritra24 replied:
Oops :p

aritra24 committed:
Undid the custom codegen changes

aritra24 committed:
Excluded the generated files from mypy

bugraoz93 commented on .pre-commit-config.yaml:
```diff
@@ -1357,6 +1357,13 @@ repos:
+        language: python
+        entry: uv run -p 3.12 --no-dev --no-progress --active --group codegen --project apache-airflow-ctl --directory airflow-ctl/ datamodel-codegen
+        pass_filenames: false
+        files: ^airflow-core/src/airflow/api_fastapi/core_api/.*\.py$
```
```suggestion
        files: ^airflow-core/src/airflow/api_fastapi/core_api/datamodels/.*\.py$
```

aritra24 replied:
done

bugraoz93 commented on .pre-commit-config.yaml:
```diff
@@ -1357,6 +1357,13 @@ repos:
+        entry: uv run -p 3.12 --no-dev --no-progress --active --group codegen --project apache-airflow-ctl --directory airflow-ctl/ datamodel-codegen
+        pass_filenames: false
+        files: ^airflow-core/src/airflow/api_fastapi/core_api/.*\.py$
+        require_serial: true
```
Could you please add the same things for auth datamodels?
```suggestion
        require_serial: true
- id: generate-airflowctl-auth-datamodels
        name: Generate Datamodels for AirflowCTL
        language: python
        entry: uv run -p 3.12 --no-dev --no-progress --active --group codegen --project apache-airflow-ctl --directory airflow-ctl/ datamodel-codegen --input=../airflow-core/src/airflow/api_fastapi/auth/managers/simple/openapi/v1-generated.yaml --output=src/airflowctl/api/datamodels/auth_generated.py
        pass_filenames: false
        files: ^airflow-core/src/airflow/api_fastapi/auth/managers/simple/datamodels/.*\.py$
        require_serial: true
```

aritra24 replied:
Added under the same pre-commit step

bugraoz93 reviewed (commented):
Looks good! Thanks Aritra!

bugraoz93 commented on airflow-ctl/pyproject.toml:
```diff
@@ -20,7 +20,7 @@ name = "apache-airflow-ctl"
 description = "Apache Airflow command line tool for communicating with an Apache Airflow, using the API."
 readme = { file = "README.md", content-type = "text/markdown" }
-requires-python = ">=3.9, <3.13"
+requires-python = ">=3.10, <3.13"
```
We need to do more than change this to lose support for `3.9`, so let's keep the `3.10+` PR separate. I created a PR for fixing the issue with ExtraLinks and excluding the entire package from MyPY for some Enum conversions. Let's rebase after that one is merged. 
#49250

aritra24 replied:
Ack

aritra24 replied:
removed this

aritra24 committed:
Reverted py3.10 change and updated change monitor files

potiuk commented on airflow-ctl/pyproject.toml:
```diff
@@ -127,7 +122,7 @@ use-standard-collections=true # list[] not List[]
 use-union-operator=true # 3.9+annotations, not `Union[]`
 
-url = "http://0.0.0.0:28080/openapi.json"
+input = "../airflow-core/src/airflow/api_fastapi/core_api/openapi/v1-generated.yaml"
```
I just renamed that one.

aritra24 replied:
oops, I'll update

aritra24 committed:
Merge branch 'main' into generate_models

aritra24 committed:
Changed name of the yaml file

aritra24 committed:
Push new generated files

bugraoz93 reviewed (commented):
Looks good! I will merge other one today which solves MyPy and other problem then we can merge this with all green :)

aritra24 committed:
Merge branch 'main' into generate_models

aritra24 committed:
Ignore generated files from add license

aritra24 committed:
Merge branch 'main' into generate_models

aritra24 committed:
regenerated generated.py

aritra24 committed:
Updated .rat-excludes for gen files

bugraoz93 reviewed (approved):
Thanks for the long waiting on dependencies and some peoblems! Looks good
